framework/include/multiapps/MultiApp.h

```diff
@@ -96,12 +96,21 @@ class MultiApp : public MooseObject, public SetupInterface, public Restartable
   virtual bool solveStep(Real dt, Real target_time, bool auto_advance = true) = 0;
 
   /**
-   * Actually advances time and causes output.
-   *
-   * If auto_advance=true was used in solveStep() then this function
-   * will do nothing.
+   * Advances the multi-apps time step which is important for dt selection.
+   * (Note this does not advance the *time*. That is done in Transient::endStep,
+   * which is called either directly from solveStep() for loose coupling cases
+   * or through finishStep() for Picard coupling cases)
+   */
+  virtual void incrementTStep() = 0;
+
+  /**
+   * Calls multi-apps executioners' endStep and postStep methods which creates output and advances
+   * time (not the time step; see incrementTStep()) among other things. This method is only called
+   * for Picard calculations and is called *after* the master application's executioner's endStep
+   * method. Consequently Picard may not work in conjunction with restart since the sub-app's time
+   * will not have been updated before the restartable data file is written.
    */
-  virtual void advanceStep() = 0;
+  virtual void finishStep() = 0;
```

Comment:
This is a pretty significant API update (since these are pure virtuals). You should probably leave the old one there which just calls the new one but also use `mooseDeprecated`. This will mean that the new API can't be pure virtual for now, but it looks like Rattlesnake is failing due to this update so you don't really have a good option. You'll just have to come back through after apps are updated and change this.